File: C++/Algorithms/Sorting/BogoSort.cpp
//C++ implementation of BogoSort by Gunjan

#include <iostream>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <vector>

using namespace std;

bool isSorted(const vector<int> &v){
    int i = v.size();
    while(--i){
        if(v[i] < v[i-1])
            return false;
    }
    return true;
}

void shuffle(vector<int> &v){
    for (size_t i = 0; i < v.size(); i++){
        swap(v[i], v[rand() % v.size()]);
    }
}

void bogoSort(vector<int> &v){
    while(!isSorted(v)){
        shuffle(v);
    }    
}

int main(){
    srand(time(NULL));

    vector<int> v = {1,3,4,6,7,9,10,2,5,8};
    bogoSort(v);

    for(auto i : v){
        cout << i << " ";
    }
    cout << endl;
}

File: Python/Algorithms/Sorting/bogo_sort.py
import random

def check_sorted(arr):
    """
    Check to see if the list is sorted in lexiographical order

    Args:
        arr (list): the list to sort

    Returns:
        boolean: True if sorted, False otherwise
    """
    curr_num = arr[0]
    for num in arr:
        if num < curr_num:
            return False
        curr_num = num
    return True

def bogo_sort(arr):
    """
    An implementation of bogosort, a very inefficient sorting algorithm with 
    an average time complexity of O(n*n!)

    Args:
        arr (list): the list to bogosort

    Returns:
        list: the inefficiently sorted with bogosort list
    """
    if not check_sorted(arr):
        random.shuffle(arr)
        return bogo_sort(arr)
    else:
        return arr

arr = [3, 1, 2]
print(f"Sorted list: {bogo_sort(arr)}")